### ISSUE TITLE
Class 'integer' not found

### ISSUE THREAD

**Issue by ryanengland** [paypal/sdk-core-php#66]
Hi guys,

Having a problem with the SDK when running a PreapprovalDetail request. Issue is with /­vendor/­paypal/­sdk-core-php/­lib/­PPMessage.php line 103: Class 'integer' not found.

Have done a raw dump of $map and get_object_vars($this) inside the init method of PPMessage.php, as follows:

````php
array(23) {
  ["responseenvelope.timestamp"]=>
  string(29) "2015-06-15T04:47:59.448-07:00"
  ["responseenvelope.ack"]=>
  string(7) "Success"
  ["responseenvelope.correlationid"]=>
  string(13) "2eccd42258c4e"
  ["responseenvelope.build"]=>
  string(8) "15743565"
  ["approved"]=>
  string(5) "false"
  ["cancelurl"]=>
  string(63) "****redacted****"
  ["curpayments"]=>
  string(1) "0"
  ["curpaymentsamount"]=>
  string(4) "0.00"
  ["curperiodattempts"]=>
  string(1) "0"
  ["currencycode"]=>
  string(3) "GBP"
  ["dateofmonth"]=>
  string(1) "0"
  ["dayofweek"]=>
  string(16) "NO_DAY_SPECIFIED"
  ["endingdate"]=>
  string(29) "2015-12-15T23:59:52.000-08:00"
  ["maxnumberofpaymentsperperiod"]=>
  string(3) "100"
  ["maxtotalamountofallpayments"]=>
  string(5) "10.00"
  ["paymentperiod"]=>
  string(19) "NO_PERIOD_SPECIFIED"
  ["pintype"]=>
  string(12) "NOT_REQUIRED"
  ["returnurl"]=>
  string(62) "****redacted****"
  ["memo"]=>
  string(7) "Testing"
  ["startingdate"]=>
  string(29) "2015-06-15T00:00:52.000-07:00"
  ["status"]=>
  string(6) "ACTIVE"
  ["feespayer"]=>
  string(6) "SENDER"
  ["displaymaxtotalamount"]=>
  string(4) "true"
}
array(29) {
  ["responseEnvelope"]=>
  NULL
  ["approved"]=>
  NULL
  ["cancelUrl"]=>
  NULL
  ["curPayments"]=>
  NULL
  ["curPaymentsAmount"]=>
  NULL
  ["curPeriodAttempts"]=>
  NULL
  ["curPeriodEndingDate"]=>
  NULL
  ["currencyCode"]=>
  NULL
  ["dateOfMonth"]=>
  NULL
  ["dayOfWeek"]=>
  NULL
  ["endingDate"]=>
  NULL
  ["maxAmountPerPayment"]=>
  NULL
  ["maxNumberOfPayments"]=>
  NULL
  ["maxNumberOfPaymentsPerPeriod"]=>
  NULL
  ["maxTotalAmountOfAllPayments"]=>
  NULL
  ["paymentPeriod"]=>
  NULL
  ["pinType"]=>
  NULL
  ["returnUrl"]=>
  NULL
  ["senderEmail"]=>
  NULL
  ["memo"]=>
  NULL
  ["startingDate"]=>
  NULL
  ["status"]=>
  NULL
  ["ipnNotificationUrl"]=>
  NULL
  ["addressList"]=>
  NULL
  ["feesPayer"]=>
  NULL
  ["displayMaxTotalAmount"]=>
  NULL
  ["sender"]=>
  NULL
  ["agreementType"]=>
  NULL
  ["error"]=>
  NULL
}
array(4) {
  ["timestamp"]=>
  string(29) "2015-06-15T04:47:59.448-07:00"
  ["ack"]=>
  string(7) "Success"
  ["correlationid"]=>
  string(13) "2eccd42258c4e"
  ["build"]=>
  string(8) "15743565"
}
array(4) {
  ["timestamp"]=>
  NULL
  ["ack"]=>
  NULL
  ["correlationId"]=>
  NULL
  ["build"]=>
  NULL
}
````

Seems to be $this->{$property} = new $type(); line that's erroring.

**Comment by jaypatel512** [comment]
Hey @ryanengland !

Would you mind adding one more debug line to our library manually to get a bit more detailed information. 

In your "vendor/paypal/sdk-core-php/lib/PPMessage.php", Can you modify this line to this:
```php
...
if (!$this->isBuiltInType($type = PPUtils::propertyType($this, $property), false) && !class_exists($type)) {
		trigger_error("Class $type of property $property of class ". get_class($this) . " with default $defaultValue not found.", E_USER_NOTICE);
		continue; // just ignore
}
...
```

I am trying to see if we could extract which property it is failing for, and that would help us easily fix this. 

Let me know if that helps. Thanks,
Jay.

**Comment by jaypatel512** [comment]
Hey @ryanengland !

Any updates on this ? Let me know if you were able to make the above change. I would be more than happy to help resolve this issue.

**Comment by ryanengland** [comment]
Hey @jaypatel512! Thanks for your help.

OK, so I have dropped the updated line in on line 73-76 of PPMessage.php. I have re-run the script again, but it is still not triggering on that line, it's triggering on line 103, where the class is actually instantiated.

So digging a little deeper, I tried a couple of different attempts:

Firstly, removed the ` && !class_exists($type)` from the if statement and this brought back the error "Class ResponseEnvelope of property responseEnvelope of class PreapprovalDetailsResponse with default not found." - as expected because PayPal's classes aren't loaded.

Next, removing the first part of the if statement, to just `if (!class_exists($type)) {` returns back `Undefined variable: type` - which is interesting as it seems that the code is falling into the else part of the if / else above. If I drop the `$type = PPUtils::propertyType($this, $property);` into the else statement, the code continues to fail with the above message, so it looks like for this part, that neither the if nor the else part in lines 63-72 is executing.

If I drop the `$type = PPUtils::propertyType($this, $property);` just below the if / else, the code then executes with the following message: `Class integer of property maxNumberOfPayments of class PreapprovalDetailsResponse with default not found.` - which I assume is what you were expecting / hoping for with the debug output.

Hope this helps!

**Comment by jaypatel512** [comment]
Aah ! Interesting. Thanks for your help debugging this. I think what I was looking for is, `Class integer of property maxNumberOfPayments of class PreapprovalDetailsResponse with default not found`. 

I think I found the root cause of the issue. I think this new patch would work out. But before I do it on github, it would be great if you could help me make sure it works great with yours too.

For the lines 62 to 67 in PPMessage.php located at https://github.com/paypal/sdk-core-php/blob/master/lib/PPMessage.php#L62-L67, replace it with this, and check again.

```php
foreach (get_object_vars($this) as $property => $defaultValue) {
			$propKey = strtolower($prefix . $property);
			if ($this->isBuiltInType(($type = PPUtils::propertyType($this, $property)))){
				$type = PPUtils::propertyType($this, $property);
				if (array_key_exists($propKey, $map)){
					$this->{$property} = urldecode($map[$propKey]);
				}
				continue; // string
```

**Comment by ryanengland** [comment]
Looks like that's worked! Nice!

**Comment by jaypatel512** [comment]
Oh great. In that case, I will go ahead and test it a bit more, and update the sdk-core-php. This might take some time,however, I am glad that you are unblocked. 

Let me know if similar issue arises again.

**Comment by ryanengland** [comment]
Hey @jaypatel512, just so we know, do you have an ETA for the patch being put in? Are we talking days / weeks / months? Cheers!

**Comment by jaypatel512** [comment]
Days/Weeks. Definitely not months. I just need to make sure all the sdk components that use this, is passing, and its a bit low priority fix right now. However, I will definitely be doing this fix soon.
